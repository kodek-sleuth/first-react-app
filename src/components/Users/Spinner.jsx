import React from 'react';

const Spinner = () => {
    return (
        <div>
            <img src="https://loading.io/spinners/rolling/lg.curve-bars-loading-indicator.gif" alt="" style={{ margin: 'auto', display: 'block', width: '50px' }}/>
        </div>
    )
}

export default Spinner;
