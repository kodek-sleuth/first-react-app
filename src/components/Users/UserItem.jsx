import React from "react"; 
import { Link } from 'react-router-dom';

const UserItem = (props) => {
  const { login, avatar_url } = props.user;
  return (
    <div>
      <div className="card">
        <img src={avatar_url} alt="" className="avatar"/>
        <h3>{login}</h3>
        <Link to={`/user/${login}`}><button className="more">More</button></Link>
      </div>
    </div>
  );
}

export default UserItem;