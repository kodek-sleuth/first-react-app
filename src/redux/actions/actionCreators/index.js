import '@babel/polyfill';
import axios from 'axios';
import { GET_LOADING, GET_USERS, GET_USER } from '../actionTypes';

const setLoading = (data) => dispatch => {
  dispatch({
    type: GET_LOADING,
    payload: data
  })
}

export const getUsers = () => async(dispatch) => {
    dispatch(setLoading(true))
    const people = await axios.get(`https://api.github.com/users?client_id=${process.env.REACT_APP_GITHUB_CLIENT_ID}&client_secret=${process.env.REACT_APP_GITHUB_CLIENT_SECRET}`);
    dispatch({
        type: GET_USERS,
        payload: people.data
    })
    dispatch(setLoading(false))
};

export const searchUsers = (text) => async(dispatch) => {
    dispatch(setLoading(true))
    const people = await axios.get(`https://api.github.com/search/users?q=${text}&client_id=${process.env.REACT_APP_GITHUB_CLIENT_ID}&client_secret=${process.env.REACT_APP_GITHUB_CLIENT_SECRET}`);
    dispatch({
        type: GET_USERS,
        payload: people.data.items
    })
    dispatch(setLoading(false))
};

export const getSingleUser = (username) => async(dispatch) => {
    dispatch(setLoading(true))
    const people = await axios.get(`https://api.github.com/users/${username}?client_id=${process.env.REACT_APP_GITHUB_CLIENT_ID}&client_secret=${process.env.REACT_APP_GITHUB_CLIENT_SECRET}`);
    dispatch({
        type: GET_USER,
        payload: people.data
    })
    dispatch(setLoading(false))
};


