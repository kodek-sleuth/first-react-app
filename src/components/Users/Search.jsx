import React, { Component } from 'react';
import { connect } from 'react-redux';
import { searchUsers } from '../../redux/actions/actionCreators';
import UserItem from './UserItem.jsx';
import Spinner from '../Users/Spinner.jsx';

export class Search extends Component {
    state = {
        text: ''
    }

    onChange = (e) => {
        console.log(this.props);
        this.setState({ text: e.target.value })
    }

    onSubmit = (e) => {
        e.preventDefault();
        this.props.searchUsers(this.state.text)
    }

    render() {
        if (this.props.loading){
            return <Spinner />
        }
        return (
            <div>
                <form style={{ marginTop: '100px', marginLeft: '100px' }} onSubmit={this.onSubmit}>
                    <input type="search" placeholder="search..." className="formControl" value={this.state.text} onChange={this.onChange} />
                    <input type="submit" value="Search" className="formControlBtn" />
                </form>
                <div className="container">
                    {
                        this.props.users.map(user => 
                        <div key={user.id}>
                            <UserItem user={user} />
                        </div>)
                    }
                </div>
            </div>
        )
    }
}

const mapStateToProps = state => ({
    loading: state.loading,
    users: state.users
})

export default connect(mapStateToProps, {searchUsers})(Search)
