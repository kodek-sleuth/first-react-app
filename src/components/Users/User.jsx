import React, { useEffect } from 'react'; 
import { connect } from 'react-redux';
import { Link, Redirect } from 'react-router-dom';
import { getSingleUser } from '../../redux/actions/actionCreators';
import Spinner from '../Users/Spinner.jsx';

const User = props => {
  useEffect(() => {
    const username = props.match.params.login;
    props.getSingleUser(username)
  }, [])

  const redirectToProfile = () => {
    window.location = props.user.blog
  }

  const { name, avatar_url } = props.user;
  if (props.loading){
    return <Spinner />
  }
  return (
    <div className="card">
      <img src={avatar_url} alt="" className="avatar"/>
      <h3>{name}</h3>
      <button className="more" onClick={redirectToProfile}>Profile</button>
    </div>
  );
}

const mapStateToProps = state => ({
  loading: state.loading,
  user: state.user
})

export default connect(mapStateToProps, {getSingleUser})(User)