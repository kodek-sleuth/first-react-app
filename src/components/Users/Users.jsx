import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { getUsers } from '../../redux/actions/actionCreators';
import UserItem from './UserItem.jsx';
import Spinner from '../Users/Spinner.jsx';

export const Users = props => {
    useEffect(() => {
        props.getUsers();
    }, [])

    if (props.loading){
        return <Spinner />
    }

    return(
        <div className="container">
            {props.users.map(user => 
                <div key={user.id}>
                    <UserItem user={user}/>
                </div>
            )}
        </div>
    )
}

const mapStateToProps = state => ({
    users: state.users,
    loading: state.loading
})

export default connect(mapStateToProps, {getUsers})(Users)
