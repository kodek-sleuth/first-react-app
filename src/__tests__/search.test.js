import React from 'react';
import moxios from 'moxios';
import axios from 'axios';
import thunk from 'redux-thunk';
import Adapter from 'enzyme-adapter-react-16';
import configureMockStore from 'redux-mock-store';
import { configure, shallow, mount } from 'enzyme';
import reducer from '../redux/reducers';
import { MemoryRouter } from 'react-dom';
import { Search } from '../components/Users/Search';
import { searchUsers } from '../redux/actions/actionCreators'

configure({ adapter: new Adapter() });

// Things to test in actions
// Mock request response

// We mock the store to save all dispatched actions on the store Instance
const createStore = configureMockStore([thunk]);
const store = createStore(reducer.initialState);

const searchProps = {
    searchUsers: jest.fn(),
    text: 'mojombo',
    loading: false,
    users: [{
        id: 1,
        login: 'mojombo',
        avatar_url: "https://avatars0.githubusercontent.com/u/1?v=4",
        blog: "https://avatars0.githubusercontent.com/u/1?v=4"
    }]
}

describe('<Search />', () => {
    beforeEach(() => {
            // moxios intercepts axios calls made to our api
            moxios.install(axios);
            moxios.stubRequest(`https://api.github.com/search/users?q=mojombo}&client_id=${process.env.REACT_APP_GITHUB_CLIENT_ID}&client_secret=${process.env.REACT_APP_GITHUB_CLIENT_SECRET}`, {
            users: [{
                id: 1,
                login: 'mojombo',
                avatar_url: "https://avatars0.githubusercontent.com/u/1?v=4",
                blog: "https://avatars0.githubusercontent.com/u/1?v=4" 
            }]
        });
        store.clearActions();
      });
      
      afterEach(() => {
        moxios.uninstall(axios);
      });
    
    it('should render a search component', () => {
        const component = shallow(<Search {...searchProps} />);
        expect(component).toMatchSnapshot();
    });

    it('should show a form', () => {
        const component = shallow(<Search {...searchProps} />);
        expect(component.find('form')).toHaveLength(1)
    });

    it('should dispatch an action to searchUsers', () => {
        store.dispatch(searchUsers());
        expect(store.getActions()).toEqual([{ type: 'GET_LOADING', payload: true }]);
    });
})