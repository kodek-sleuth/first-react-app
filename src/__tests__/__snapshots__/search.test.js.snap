// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> should render a search component 1`] = `ShallowWrapper {}`;
